### PR TITLE
[14.0][ADD] product_variant_company

### PR DESCRIPTION
**Pull request by Volodiay622**
on this [comment](https://github.com/OCA/stock-logistics-barcode/pull/526#discussion_r1275978467) from @[rousseldenis](https://github.com/rousseldenis)

### CODE REVIEW

File: product_company_id/readme/CONTRIBUTORS.rst [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Sylvain LE GAL <https://twitter.com/legalsylvain>
```
Comment (legalsylvain) [inline product_company_id/readme/CONTRIBUTORS.rst]: This module comes from a module I développed ?

**Review by legalsylvain — COMMENTED** [review]
Do we have to add a new rule to target the new company field ?

File: product_company_id/readme/CONTRIBUTORS.rst (hunk 1 above)
Comment (Volodiay622) [reply product_company_id/readme/CONTRIBUTORS.rst]: it's from [this PR](https://github.com/OCA/stock-logistics-barcode/pull/526)

**Comment by francesco-ooops** [comment]
ping @simahawk @rousseldenis in relation to [#526](https://github.com/OCA/stock-logistics-barcode/pull/526)

thanks!

**Comment by rousseldenis** [comment]
> Do we have to add a new rule to target the new company field ?

@legalsylvain What do you mean ?

**Comment by rousseldenis** [comment]
> > Do we have to add a new rule to target the new company field ?
> 
> @legalsylvain What do you mean ?

Oh, ir.rule. Yes, indeed. @Volodiay622 Could you do it ?

**Comment by legalsylvain** [comment]
> @legalsylvain What do you mean ?
> Oh, ir.rule. Yes, indeed. @Volodiay622 Could you do it ?

Well, in fact, I don't know how odoo works ! I _guess_ that an ir.rule is applied, when searching on ``product.procuct`` but I don't see ir.rule related to product.product model. There is one, related to product.template. 
https://github.com/odoo/odoo/blob/14.0/addons/product/security/product_security.xml#L34

I so _guess_ that this rule is applied, but making a join that could cost time. So I _gess_ that we should add a new rule on product.product, based on the new company field. But in fact, I'm not sure it will improve something. Maybe two WHEN will be added in the SQL request...
What do you think ?

File: product_company_id/models/product_product.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+from odoo import fields, models
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    # Store related field
+    company_id = fields.Many2one(
+        related="product_tmpl_id.company_id", store=True, readonly=False
```
Comment (legalsylvain) [inline product_company_id/models/product_product.py]:
```suggestion
        related="product_tmpl_id.company_id", store=True, readonly=False, index=True
```

Comment (Volodiay622) [reply product_company_id/models/product_product.py]: @legalsylvain I added your suggested change

**Comment by francesco-ooops** [comment]
@legalsylvain @rousseldenis are we good to go?

File: product_company_id/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+{
+    "name": "Product Company ID",
```
Comment (simahawk) [inline product_company_id/__manifest__.py]: why not product_company only? I mean, why not removing "id" everywhere?
Not a blocker!

File: product_company_id/tests/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+from . import test_module
```
Comment (simahawk) [inline product_company_id/tests/__init__.py]:
```suggestion
from . import test_product_company
```

File: product_company_id/tests/test_module.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+from odoo.tests.common import TransactionCase
+
+
+class TestModule(TransactionCase):
+    def setUp(self):
+        super().setUp()
+
+        self.ProductProduct = self.env["product.product"]
+        self.company_1 = self.env["res.company"].create({"name": "Company 1"})
+        self.company_2 = self.env["res.company"].create({"name": "Company 2"})
```
Comment (simahawk) [inline product_company_id/tests/test_module.py]:
```suggestion
    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
        cls.ProductProduct = cls.env["product.product"]
        cls.company_1 = cls.env["res.company"].create({"name": "Company 1"})
        cls.company_2 = cls.env["res.company"].create({"name": "Company 2"})
```

File: product_company_id/__manifest__.py (hunk 3 above)
Comment (Volodiay622) [reply product_company_id/__manifest__.py]: applied

File: product_company_id/tests/__init__.py (hunk 4 above)
Comment (Volodiay622) [reply product_company_id/tests/__init__.py]: applied

File: product_company_id/tests/test_module.py (hunk 5 above)
Comment (Volodiay622) [reply product_company_id/tests/test_module.py]: applied

**Comment by francesco-ooops** [comment]
@rousseldenis @simahawk could you also provide a feedback on module naming and description?

**Comment by francesco-ooops** [comment]
@simahawk ping

**Comment by simahawk** [comment]
LGTM. Can we squash everything onto "[ADD] product_company"?

**Comment by simahawk** [comment]
> LGTM. Can we squash everything onto "[ADD] product_company"?

Thinking of it twice... Sorry... :grin: Maybe `product_variant_company` makes more sense because by default the product has already a company via its template. No?

**Comment by simahawk** [comment]
@Volodiay622 can you squash? 
@rousseldenis @legalsylvain fine for you?

**Comment by francesco-ooops** [comment]
@simahawk from description I still don't understand what this module does. 

Is it a technical or a functional module? 

For which use case would this be employed?

**Comment by simahawk** [comment]
> @simahawk from description I still don't understand what this module does.
> 
> Is it a technical or a functional module?
> 
> For which use case would this be employed?

It's a technical module. You need a company at variant level? Use this.
The description should be adapted if is not clear.

**Comment by francesco-ooops** [comment]
@simahawk ok, but functionally I really don't understand it. 

Here's a video of a new product template with two variants: when setting a company in a variant, it is changed for all variants and for product template as well

https://recordit.co/ljowC1XRUx

**Comment by simahawk** [comment]
> @simahawk ok, but functionally I really don't understand it.
> 
> Here's a video of a new product template with two variants: when setting a company in a variant, it is changed for all variants and for product template as well
> 
> https://recordit.co/ljowC1XRUx

In fact, as it is it does not add anything if not a field that could be overridden.
IMO it could be already using a computed / writeable field instead of a related one and default to the tmpl one...
@rousseldenis as you suggested this split, what did you have in mind regarding this?

**Comment by legalsylvain** [comment]
> Here's a video of a new product template with two variants: when setting a company in a variant, it is changed for all variants and for product template as well

IMO, it is the expected behaviour. 

As said in this module, it is a just technical module to bring company_id on product_product model. Nothing else. Don't you think ?

**Comment by francesco-ooops** [comment]
> > Here's a video of a new product template with two variants: when setting a company in a variant, it is changed for all variants and for product template as well
> 
> IMO, it is the expected behaviour.
> 
> As said in this module, it is a just technical module to bring company_id on product_product model. Nothing else. Don't you think ?

On a second thought, yes, it's standard behavior indeed; 

it's just that from a ux standpoint, odoo makes it confusing when accessing a variant from the "product variant" menu to understand which fields are at product template level and which fields are at variant level.

**Comment by francesco-ooops** [comment]
@legalsylvain @rousseldenis good to go?

File: product_variant_company/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
```
Comment (yostashiro) [inline product_variant_company/__manifest__.py]: Nitpicking. Same for other .py files.
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

Comment (Volodiay622) [reply product_variant_company/__manifest__.py]: resolved

**Comment by francesco-ooops** [comment]
@legalsylvain @rousseldenis ping :)

**Review by legalsylvain — APPROVED** [review]

File: product_variant_company/__manifest__.py (hunk 6 above)
Comment (francesco-ooops) [reply product_variant_company/__manifest__.py]: @yostashiro good for you?

File: product_variant_company/readme/DESCRIPTION.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+This module extend Odoo functionalities of the "product.product" model
+by adding a related field that provides information about
+the associated company for each product.
```
Comment (simahawk) [inline product_variant_company/readme/DESCRIPTION.rst]: One last thing: can you please add here the explanation that is a technical module and does not change anything in the std behavior (see related comment)?

**Review by simahawk — APPROVED** [review]
Pre-approving

**Pull request merged**